### PR TITLE
Align dictionary key converter/metadata retrieval with pattern for collection elements

### PR DESCRIPTION
**Pull request by layomia**
Alternative approach to https://github.com/dotnet/runtime/pull/48232. Here we create a `KeyClassInfo` for dictionary keys, similar to how it is done for dictionary and collection elements/values.

This is needed so incoming STJ source generator can pre-generate metadata for dictionary key types, and pass that directly to the serializer. This is analogous to the design for collection elements e.g. [here](https://github.com/dotnet/runtimelab/blob/6d9a932355cb716e1f3d22e21f64dc3385621679/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/KnownCollectionTypeInfos.cs#L20). It should also help with other scenarios described in #48232.

The breaking change described in https://github.com/dotnet/runtime/pull/48232#issuecomment-779567299 where custom converters will throw `InvalidOperationExeption` (now `NotSupportedException`, following this change) can be mitigated by exposing the functionality to handle dictionary keys: https://github.com/dotnet/runtime/issues/46520.

### CODE REVIEW

**Comment by steveharter** [comment]
Are we OK with "breaking" this and taking a hard dependency on https://github.com/dotnet/runtime/issues/46520?

**Comment by layomia** [comment]
> Are we OK with "breaking" this and taking a hard dependency on #46520?

Yes, I believe so. It is rare for users to provide custom simple converters (aside from `TimeSpan`, which will have built-in support in 6.0), especially now that we have the feature to read quoted number values. I think very few folks will hit this, and #46520 will be a good mitigation. If we take this change, I'll prioritize getting that feature in.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 1]
```diff
@@ -540,10 +542,13 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
...
+        {
+            ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
+            return default;
+        }
 
         internal virtual void WriteWithQuotes(Utf8JsonWriter writer, [DisallowNull] T value, JsonSerializerOptions options, ref WriteStack state)
-            => throw new InvalidOperationException();
+            => ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
 
         internal sealed override void WriteWithQuotesAsObject(Utf8JsonWriter writer, object value, JsonSerializerOptions options, ref WriteStack state)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:553]: Why do we box here?

**Review by steveharter — APPROVED** [review]
Looks good; we may need to react if anyone with a primitive custom converter also needs quoted number support.

However, even before this PR that scenario could be considered broken since their custom converter wasn't used for quoted numbers.

File: src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.Dictionary.NonStringKey.cs [hunk 2]
```diff
@@ -619,5 +619,70 @@ void RunTest<T>(T dictionary)
...
+                exAsStr = ex.ToString();
+                Assert.Contains(keyTypeAsStr, exAsStr);
+                Assert.Contains(converterTypeAsStr, exAsStr);
+            }
+        }
+
+        private class ClassWithNonStringDictKeys
+        {
+            public Dictionary<int, string> PrimitiveKey { get; set; }
+
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.Dictionary.NonStringKey.cs:676]: nit: unnecessary whitespace?

File: src/libraries/System.Text.Json/src/Resources/Strings.resx [hunk 3]
```diff
@@ -554,4 +554,4 @@
   <data name="DeserializeUnableToAssignNull" xml:space="preserve">
     <value>Unable to assign 'null' to the property or field of type '{0}'.</value>
   </data>
-</root>
+</root>
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/Resources/Strings.resx:557]: Github is complaining about no newline at EOF.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:553]: The only caller of this specific method is from `ObjectConverter`, when serializing `System.Object`-typed keys. The values to serialize are boxed at this point. `ObjectConverter` basically forwards the (boxed) value to the typed converter for the runtime type of the instance. The typed converter then unboxes the value and serializes it. 

It's the same as how `TryWriteAsObject` works.

**Comment by layomia** [comment]
The breaking change here was reversed in https://github.com/dotnet/runtime/pull/57138 and https://github.com/dotnet/runtime/pull/57302.
